File: ios-native/GatherLists/GatherLists/Models/RecipeAttributes.swift
import Foundation

/// A labelled value in one of the recipe attribute vocabularies.
protocol RecipeAttributeOption: RawRepresentable, CaseIterable, Identifiable, Hashable where RawValue == String {
    var label: String { get }
}

extension RecipeAttributeOption {
    var id: String { rawValue }
}

enum RecipeCourse: String, RecipeAttributeOption {
    case main, side, dessert, snack, drink, component

    var label: String {
        switch self {
        case .main: return "Main dish"
        case .side: return "Side"
        case .dessert: return "Dessert"
        case .snack: return "Snack"
        case .drink: return "Drink"
        case .component: return "Sauce / base"
        }
    }
}

enum RecipeProtein: String, RecipeAttributeOption {
    case chicken, beef, pork, lamb, turkey, fish, shellfish, eggs, tofu, beans, dairy, none

    var label: String {
        switch self {
        case .tofu: return "Tofu / tempeh"
        case .beans: return "Beans / lentils"
        case .dairy: return "Cheese / dairy"
        case .none: return "No main protein"
        default: return rawValue.capitalized
        }
    }
}

enum RecipeCuisine: String, RecipeAttributeOption {
    case american, mexican, italian, french, spanish, greek, mediterranean
    case middleEastern = "middle_eastern"
    case indian, chinese, japanese, korean, thai, vietnamese, caribbean, african, other

    var label: String {
        switch self {
        case .middleEastern: return "Middle Eastern"
        default: return rawValue.capitalized
        }
    }
}

enum RecipeEffort: String, RecipeAttributeOption {
    case quick, medium, project

    var label: String {
        switch self {
        case .quick: return "Quick (under 30 min)"
        case .medium: return "Medium"
        case .project: return "Project (1 hr+)"
        }
    }
}

enum RecipeMethod: String, RecipeAttributeOption {
    case stovetop, oven, grill
    case slowCooker = "slow_cooker"
    case pressureCooker = "pressure_cooker"
    case airFryer = "air_fryer"
    case fried
    case noCook = "no_cook"

    var label: String {
        switch self {
        case .stovetop: return "Stovetop"
        case .oven: return "Oven"
        case .grill: return "Grill"
        case .slowCooker: return "Slow cooker"
        case .pressureCooker: return "Pressure cooker"
        case .airFryer: return "Air fryer"
        case .fried: return "Fried"
        case .noCook: return "No cook"
        }
    }
}

/// Planner-facing attributes of a recipe, mapped to the `recipe_attributes` Supabase table.
/// Categorical values are stored as raw strings so a value added in a later build doesn't
/// break decoding here; use the typed accessors.
struct RecipeAttributes: Codable, Hashable {
    let recipeId: UUID
    var course: String?
    var mealTypes: [String]
    var protein: String?
    var cuisine: String?
    var effort: String?
    var method: String?
    var kidFriendly: Bool?
    var perishables: [String]
    var manualFields: [String]
    var autoSourceHash: String?
    var autoTaggedAt: Date?

    /// Field names used in `manualFields`.
    enum Field: String, CaseIterable {
        case course
        case mealTypes = "meal_types"
        case protein, cuisine, effort, method
        case kidFriendly = "kid_friendly"
    }

    init(recipeId: UUID) {
        self.recipeId = recipeId
        mealTypes = []
        perishables = []
        manualFields = []
    }

    var courseValue: RecipeCourse? { course.flatMap(RecipeCourse.init(rawValue:)) }
    var proteinValue: RecipeProtein? { protein.flatMap(RecipeProtein.init(rawValue:)) }
    var cuisineValue: RecipeCuisine? { cuisine.flatMap(RecipeCuisine.init(rawValue:)) }
    var effortValue: RecipeEffort? { effort.flatMap(RecipeEffort.init(rawValue:)) }
    var methodValue: RecipeMethod? { method.flatMap(RecipeMethod.init(rawValue:)) }
    var mealTypeValues: [MealType] { MealType.allCases.filter { mealTypes.contains($0.rawValue) } }

    func isManual(_ field: Field) -> Bool {
        manualFields.contains(field.rawValue)
    }

    /// Short labels for display chips, in a stable order.
    var displayChips: [String] {
        var chips = mealTypeValues.map(\.label)
        if let courseValue, courseValue != .main { chips.append(courseValue.label) }
        if let proteinValue, proteinValue != .none { chips.append(proteinValue.label) }
        if let cuisineValue, cuisineValue != .other { chips.append(cuisineValue.label) }
        if let effortValue { chips.append(effortValue == .quick ? "Quick" : effortValue == .project ? "Project" : "Medium effort") }
        if let methodValue { chips.append(methodValue.label) }
        if kidFriendly == true { chips.append("Kid-friendly") }
        return chips
    }

    enum CodingKeys: String, CodingKey {
        case recipeId = "recipe_id"
        case course
        case mealTypes = "meal_types"
        case protein, cuisine, effort, method
        case kidFriendly = "kid_friendly"
        case perishables
        case manualFields = "manual_fields"
        case autoSourceHash = "auto_source_hash"
        case autoTaggedAt = "auto_tagged_at"
    }
}

File: src/utils/recipeAttributes.js
/**
 * Recipe attribute vocabulary shared by the web recipe details UI (and, later,
 * the planner). Values match the `recipe_attributes` check constraints and the
 * iOS enums in RecipeAttributes.swift.
 */

export const COURSES = [
  { id: 'main', label: 'Main dish' },
  { id: 'side', label: 'Side' },
  { id: 'dessert', label: 'Dessert' },
  { id: 'snack', label: 'Snack' },
  { id: 'drink', label: 'Drink' },
  { id: 'component', label: 'Sauce / base' },
];

export const PROTEINS = [
  { id: 'chicken', label: 'Chicken' },
  { id: 'beef', label: 'Beef' },
  { id: 'pork', label: 'Pork' },
  { id: 'lamb', label: 'Lamb' },
  { id: 'turkey', label: 'Turkey' },
  { id: 'fish', label: 'Fish' },
  { id: 'shellfish', label: 'Shellfish' },
  { id: 'eggs', label: 'Eggs' },
  { id: 'tofu', label: 'Tofu / tempeh' },
  { id: 'beans', label: 'Beans / lentils' },
  { id: 'dairy', label: 'Cheese / dairy' },
  { id: 'none', label: 'No main protein' },
];

export const CUISINES = [
  'american', 'mexican', 'italian', 'french', 'spanish', 'greek', 'mediterranean',
  'middle_eastern', 'indian', 'chinese', 'japanese', 'korean', 'thai',
  'vietnamese', 'caribbean', 'african', 'other',
].map((id) => ({
  id,
  label: id === 'middle_eastern' ? 'Middle Eastern' : id.charAt(0).toUpperCase() + id.slice(1),
}));

export const EFFORTS = [
  { id: 'quick', label: 'Quick (under 30 min)', chip: 'Quick' },
  { id: 'medium', label: 'Medium', chip: 'Medium effort' },
  { id: 'project', label: 'Project (1 hr+)', chip: 'Project' },
];

export const METHODS = [
  { id: 'stovetop', label: 'Stovetop' },
  { id: 'oven', label: 'Oven' },
  { id: 'grill', label: 'Grill' },
  { id: 'slow_cooker', label: 'Slow cooker' },
  { id: 'pressure_cooker', label: 'Pressure cooker' },
  { id: 'air_fryer', label: 'Air fryer' },
  { id: 'fried', label: 'Fried' },
  { id: 'no_cook', label: 'No cook' },
];

/** Field names recorded in `manualFields` when a person edits them. */
export const ATTRIBUTE_FIELDS = ['course', 'mealTypes', 'protein', 'cuisine', 'effort', 'method', 'kidFriendly'];

/** DB column name for each field, as stored in `manual_fields` (shared with iOS). */
export const FIELD_COLUMNS = {
  course: 'course',
  mealTypes: 'meal_types',
  protein: 'protein',
  cuisine: 'cuisine',
  effort: 'effort',
  method: 'method',
  kidFriendly: 'kid_friendly',
};

const labelOf = (options, id) => options.find((o) => o.id === id)?.label;

/**
 * An empty attributes object for a recipe with none yet.
 * @param {string} recipeId
 */
export const emptyAttributes = (recipeId) => ({
  recipeId,
  course: null,
  mealTypes: [],
  protein: null,
  cuisine: null,
  effort: null,
  method: null,
  kidFriendly: null,
  perishables: [],
  manualFields: [],
  autoSourceHash: null,
  autoTaggedAt: null,
});

/**
 * Short labels for display chips, in the same order as iOS.
 * @param {object|null} attributes
 * @returns {string[]}
 */
export const attributeChips = (attributes) => {
  if (!attributes) return [];
  const chips = ['breakfast', 'lunch', 'dinner']
    .filter((m) => attributes.mealTypes?.includes(m))
    .map((m) => m.charAt(0).toUpperCase() + m.slice(1));
  if (attributes.course && attributes.course !== 'main') chips.push(labelOf(COURSES, attributes.course));
  if (attributes.protein && attributes.protein !== 'none') chips.push(labelOf(PROTEINS, attributes.protein));
  if (attributes.cuisine && attributes.cuisine !== 'other') chips.push(labelOf(CUISINES, attributes.cuisine));
  if (attributes.effort) chips.push(EFFORTS.find((e) => e.id === attributes.effort)?.chip);
  if (attributes.method) chips.push(labelOf(METHODS, attributes.method));
  if (attributes.kidFriendly === true) chips.push('Kid-friendly');
  return chips.filter(Boolean);
};

/**
 * Returns `draft` with `manualFields` extended by every field that differs from `original`,
 * so automatic tagging on iOS never overwrites a person's choice.
 * @param {object} original
 * @param {object} draft
 * @returns {object}
 */
export const markManualEdits = (original, draft) => {
  const manual = new Set(original.manualFields ?? []);
  for (const field of ATTRIBUTE_FIELDS) {
    const before = original[field];
    const after = draft[field];
    const changed = Array.isArray(before) || Array.isArray(after)
      ? JSON.stringify(before ?? []) !== JSON.stringify(after ?? [])
      : (before ?? null) !== (after ?? null);
    if (changed) manual.add(FIELD_COLUMNS[field]);
  }
  return {
    ...draft,
    manualFields: ATTRIBUTE_FIELDS.map((f) => FIELD_COLUMNS[f]).filter((c) => manual.has(c)),
  };
};
